In [1]:
# Knock1: Data を読み込んでみよう

import pandas as pd

customer_master = pd.read_csv('sample_code/chapter_1/customer_master.csv')
customer_master.head()

customer_id customer_name    registration_date customer_name_kana  \
0    IK152942        平田 裕次郎  2019-01-01 00:25:33          ひらた ゆうじろう   
1    TS808488         田村 詩織  2019-01-01 01:13:45            たむら しおり   
2    AS834628         久野 由樹  2019-01-01 02:00:14             ひさの ゆき   
3    AS345469          鶴岡 薫  2019-01-01 04:48:22           つるおか かおる   
4    GD892565         大内 高史  2019-01-01 04:54:51           おおうち たかし   

                         email gender  age      birth pref  
0  hirata_yuujirou@example.com      M   29  1990/6/10  石川県  
1    tamura_shiori@example.com      F   33  1986/5/20  東京都  
2      hisano_yuki@example.com      F   63   1956/1/2  茨城県  
3   tsuruoka_kaoru@example.com      M   74  1945/3/25  東京都  
4   oouchi_takashi@example.com      M   54   1965/8/5  千葉県

In [2]:
item_master = pd.read_csv('sample_code/chapter_1/item_master.csv')
item_master.head()

item_id item_name  item_price
0    S001      PC-A       50000
1    S002      PC-B       85000
2    S003      PC-C      120000
3    S004      PC-D      180000
4    S005      PC-E      210000

In [3]:
transaction_1 = pd.read_csv('sample_code/chapter_1/transaction_1.csv')
transaction_1.head()

transaction_id   price         payment_date customer_id
0    T0000000113  210000  2019-02-01 01:36:57    PL563502
1    T0000000114   50000  2019-02-01 01:37:23    HD678019
2    T0000000115  120000  2019-02-01 02:34:19    HD298120
3    T0000000116  210000  2019-02-01 02:47:23    IK452215
4    T0000000117  170000  2019-02-01 04:33:46    PL542865

In [4]:
transaction_detail_1 = pd.read_csv('sample_code/chapter_1/transaction_detail_1.csv')
transaction_detail_1.head()

detail_id transaction_id item_id  quantity
0          0    T0000000113    S005         1
1          1    T0000000114    S001         1
2          2    T0000000115    S003         1
3          3    T0000000116    S005         1
4          4    T0000000117    S002         2

### Data の大枠
- customer_master: 顧客の性別や年齢などの顧客詳細情報。
- item_master: 商品名や商品単価の情報。
- transaction: いつ誰が購入したかという情報。
- transaction_detail: 購入した商品や数量などの情報。

売上とは切っても切り離せない最も粒度が細かい売上関連の Data である **transaction_detail** を主軸に

1. transaction_detail_1, 2や、transaction_1, 2を縦に結合する。(Union)
2. transaction_detail をもとに transaction, customer_master, item_master を横に結合する。(Join)

In [5]:
# Knock2: Data の Union
transaction_2 = pd.read_csv('sample_code/chapter_1/transaction_2.csv')
transaction = pd.concat([transaction_1, transaction_2], ignore_index=True)
display(transaction.head())

print(len(transaction_1))
print(len(transaction_2))
print(len(transaction))

transaction_detail_2 = pd.read_csv('sample_code/chapter_1/transaction_detail_2.csv')
transaction_detail = pd.concat([transaction_detail_1, transaction_detail_2], ignore_index=True)
display(transaction_detail.head())

print(len(transaction_detail_1))
print(len(transaction_detail_2))
print(len(transaction_detail))

transaction_id   price         payment_date customer_id
0    T0000000113  210000  2019-02-01 01:36:57    PL563502
1    T0000000114   50000  2019-02-01 01:37:23    HD678019
2    T0000000115  120000  2019-02-01 02:34:19    HD298120
3    T0000000116  210000  2019-02-01 02:47:23    IK452215
4    T0000000117  170000  2019-02-01 04:33:46    PL542865

5000
1786
6786


detail_id transaction_id item_id  quantity
0          0    T0000000113    S005         1
1          1    T0000000114    S001         1
2          2    T0000000115    S003         1
3          3    T0000000116    S005         1
4          4    T0000000117    S002         2

5000
2144
7144


In [6]:
# Knock3: Transaction data の Join
join_data = pd.merge(transaction_detail, transaction[['transaction_id', 'payment_date', 'customer_id']],
                     on='transaction_id', how='left')
display(join_data.head())

print(len(transaction_detail))
print(len(transaction))
print(len(join_data))

detail_id transaction_id item_id  quantity         payment_date customer_id
0          0    T0000000113    S005         1  2019-02-01 01:36:57    PL563502
1          1    T0000000114    S001         1  2019-02-01 01:37:23    HD678019
2          2    T0000000115    S003         1  2019-02-01 02:34:19    HD298120
3          3    T0000000116    S005         1  2019-02-01 02:47:23    IK452215
4          4    T0000000117    S002         2  2019-02-01 04:33:46    PL542865

7144
6786
7144


In [7]:
# Knock4: Master data の Join
join_data = pd.merge(join_data, customer_master, on='customer_id', how='left')
join_data = pd.merge(join_data, item_master, on='item_id', how='left')
join_data.head()

detail_id transaction_id item_id  quantity         payment_date  \
0          0    T0000000113    S005         1  2019-02-01 01:36:57   
1          1    T0000000114    S001         1  2019-02-01 01:37:23   
2          2    T0000000115    S003         1  2019-02-01 02:34:19   
3          3    T0000000116    S005         1  2019-02-01 02:47:23   
4          4    T0000000117    S002         2  2019-02-01 04:33:46   

  customer_id customer_name    registration_date customer_name_kana  \
0    PL563502         井本 芳正  2019-01-07 14:34:35           いもと よしまさ   
1    HD678019         三船 六郎  2019-01-27 18:00:11           みふね ろくろう   
2    HD298120         山根 小雁  2019-01-11 08:16:02            やまね こがん   
3    IK452215         池田 菜摘  2019-01-10 05:07:38            いけだ なつみ   
4    PL542865         栗田 憲一  2019-01-25 06:46:05           くりた けんいち   

                         email gender  age       birth pref item_name  \
0  imoto_yoshimasa@example.com      M   30   1989/7/15  熊本県      PC-E   
1   mifune_rokurou@example.com      M   73  1945/11/29  京都府      PC-A   
2     yamane_kogan@example.com      M   42   1977/5/17  茨城県      PC-C   
3    ikeda_natsumi@example.com      F   47   1972/3/17  兵庫県      PC-E   
4   kurita_kenichi@example.com      M   74  1944/12/17  長崎県      PC-B   

   item_price  
0      210000  
1       50000  
2      120000  
3      210000  
4       85000

結合した影響で売上（ Price ）が落ちているので、計算する必要あり。

In [8]:
# Knock5: 必要な Data 列を作ろう
join_data['price'] = join_data['quantity'] * join_data['item_price']
join_data[['quantity', 'item_price', 'price']].head()

quantity  item_price   price
0         1      210000  210000
1         1       50000   50000
2         1      120000  120000
3         1      210000  210000
4         2       85000  170000

1. Data 加工前お transaction data における price
2. Data 加工後に計算によって作成した price 列

上記は、合計すると同じ値になるはずなので検算を行う。

In [9]:
# Knock6 Data 検算をしよう
print(join_data['price'].sum())
print(transaction['price'].sum())

join_data['price'].sum() == transaction['price'].sum()

971135000
971135000


True

In [10]:
# Knock7: 各種統計量を把握しよう
join_data.isnull().sum()

detail_id             0
transaction_id        0
item_id               0
quantity              0
payment_date          0
customer_id           0
customer_name         0
registration_date     0
customer_name_kana    0
email                 0
gender                0
age                   0
birth                 0
pref                  0
item_name             0
item_price            0
price                 0
dtype: int64

In [11]:
join_data.describe()

detail_id     quantity          age     item_price          price
count  7144.000000  7144.000000  7144.000000    7144.000000    7144.000000
mean   3571.500000     1.199888    50.265677  121698.628219  135937.150056
std    2062.439494     0.513647    17.190314   64571.311830   68511.453297
min       0.000000     1.000000    20.000000   50000.000000   50000.000000
25%    1785.750000     1.000000    36.000000   50000.000000   85000.000000
50%    3571.500000     1.000000    50.000000  102500.000000  120000.000000
75%    5357.250000     1.000000    65.000000  187500.000000  210000.000000
max    7143.000000     4.000000    80.000000  210000.000000  420000.000000

price から読み取れるもの
- 平均: 135,937円
- 最高金額: 420,000円
    単価210,000円の PC を２台買っている User がいると想像できる。

quantity から読み取れるもの
- 最高は「４」だが 75% の中でも「１」なので、ほとんどの顧客が１で購入している。

age から読み取れるもの
- 20歳から80歳までの範囲の顧客層

In [12]:
print(join_data['payment_date'].min())
print(join_data['payment_date'].max())

2019-02-01 01:36:57
2019-07-31 23:41:38


Data の期間範囲を追加で確認しておく。

In [13]:
# 月別で Data を集計してみよう

detail_id              int64
transaction_id        object
item_id               object
quantity               int64
payment_date          object
customer_id           object
customer_name         object
registration_date     object
customer_name_kana    object
email                 object
gender                object
age                    int64
birth                 object
pref                  object
item_name             object
item_price             int64
price                  int64
dtype: object

月別に集計して一覧表示してみる
1. 購入日である payment_date から年月の列を作成する
2. 年月単位で price を集計し、表示する。

In [ ]:
# payment_data の Data 型を確認。
join_data.dtypes

In [14]:
# payment_data が object 型。
# 今後も踏まえて datetime 型に変更して年月列を作成。
join_data['payment_date'] = pd.to_datetime(join_data['payment_date'])
join_data['payment_month'] = join_data['payment_date'].dt.strftime('%Y%m')
join_data[['payment_date', 'payment_month']].head()

payment_date payment_month
0 2019-02-01 01:36:57        201902
1 2019-02-01 01:37:23        201902
2 2019-02-01 02:34:19        201902
3 2019-02-01 02:47:23        201902
4 2019-02-01 04:33:46        201902

In [15]:
# group_by でまとめたい列と、集計方法を指定。
# price のみを表示させる為、最後に price 列を指定。
join_data.groupby('payment_month').sum()['price']

payment_month
201902    160185000
201903    160370000
201904    160510000
201905    155420000
201906    164030000
201907    170620000
Name: price, dtype: int64

- ５月が減少したが、６，７月で回復。
- ７月が半年間で最も売上が高い。
- 約１億６千万くらいの単月売上で年間で20億くらいの売上が期待。